File: Java/Search_a_2D_Matrix_II.java
/**
 * Problem: https://leetcode.com/problems/search-a-2d-matrix-ii/
 * 
 * Search from bottom left to top right. 
 * Since strictly ascending in any row or column,
 * use current position(r, c) as the bottom right boundary of the matrix search.
 * 
 */

public class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        int n = matrix[0].length;
        
        int r = m-1;
        int c = 0;
        while(r >= 0 && c < n) {
            if(matrix[r][c] == target) return true;
            else if(target > matrix[r][c]) {
                c++;
            } else {
                r--;
            }
        }
        return false;
    }
}

File: Swift/search-a-2d-matrix-ii.swift
/**
 * https://leetcode.com/problems/search-a-2d-matrix-ii/
 * 
 * 
 */ 
// Date: Tue Feb 23 08:21:00 PST 2021
class Solution {
    func searchMatrix(_ matrix: [[Int]], _ target: Int) -> Bool {
        if matrix.count == 0 || matrix[0].count == 0 { return false }
        var row = matrix.count - 1
        var col = 0
        while col < matrix[0].count, row >= 0 {
            if matrix[row][col] == target { return true }
            if matrix[row][col] > target { row -= 1 }
            else { col += 1 }
        }
        return false
    }
}